File: src/ranking/ranking.ts
/**
 * Ranking logic
 *
 * ASSUMES VALID input ;; does not fetch data
 * 
 * MINDFUL: location, budget, time; 
 */


// location
export type LatLng = {
  lat: number;
  lng: number;
};


export type UserPreferences = {
  budgetRangePreference: "low" | "medium" | "high";
  walkingDistancePreference: "close" | "medium" | "far";
  stayTimePreference: "short" | "medium" | "long";
};

export type ParkingSpot = { // parking spot == candidate meter in data layer
  spaceid: string;
  latlng: LatLng;
  rate: number;        // $ per hour from raterange 
  timelimit: number;   // minutes from 2hr
};

export type RankedParkingSpot = ParkingSpot & { // ranked parking spot == ranked meter in data layer
  distanceToDestination: number; // in meters
  walkTime: number;              // in minutes
  estimatedTotalCost: number;    // cost 
  score: number;
  rank: number;
};


// helper functions 
/**
 * this will compute distance between two lat/lng points in meters -- aka haversine
 */
function distanceMeters(a: LatLng, b: LatLng): number {
    const R = 6371000;
    const toRad = (d: number) => (d * Math.PI) / 180;

    const dLat = toRad(b.lat - a.lat);
    const dLng = toRad(b.lng - a.lng);
    const lat1 = toRad(a.lat);
    const lat2 = toRad(b.lat);

    const h =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLng / 2) ** 2;

  return 2 * R * Math.asin(Math.sqrt(h));
}

/**
 * this converts user preferences into scoring weights.
 */
function getWeights(prefs: UserPreferences) {
  return {
    distance: prefs.walkingDistancePreference === "close" ? 0.45 : 0.3,
    cost: prefs.budgetRangePreference === "low" ? 0.45 : 0.3,
    time: prefs.stayTimePreference === "long" ? 0.25 : 0.2,
  };
}


// main logic: //
/**
 * Rank candidate parking spots based on context-aware scoring.
 */
export function rankParkingSpots(
  candidates: ParkingSpot[],
  destination: LatLng,
  userPrefs: UserPreferences
): RankedParkingSpot[] {

  const weights = getWeights(userPrefs);

  // init bounds
  const MAX_DISTANCE = 1000;
  const MAX_COST = 20;   
  const IDEAL_TIME = 180; 
  
  const scored = candidates.map((spot) => {
  const distance = distanceMeters(spot.latlng, destination);
  const walkTime = Math.ceil(distance / 80); // avg 80 m/min
  const estimatedTotalCost = spot.rate * (spot.timelimit / 60);

  // TODO: normalize distance, cost, and time scores from 0 to 1
  const distanceScore = 1 - Math.min(distance / MAX_DISTANCE, 1);
  const costScore = 1- Math.min(estimatedTotalCost / MAX_COST, 1);
  const timeScore = Math.min(spot.timelimit / IDEAL_TIME, 1);

  const score =
    weights.distance * distanceScore +
    weights.cost * costScore +
    weights.time * timeScore;

  return {
    ...spot,
    distanceToDestination: distance,
    walkTime,
    estimatedTotalCost,
    score,
    rank: -1, // marked after sorting
  };
  });


  // sorts by score
  scored.sort((a, b) => b.score - a.score);

  // assign ranks and return top K rankings
  return scored.slice(0, 8).map((spot, index) => ({
    ...spot,
    rank: index + 1,
  }));
}

File: src/ranking/ranking.py
"""
Ranking logic

ASSUMES VALID input ;; does not fetch data

MINDFUL: location, budget, time;
"""

from math import sin, cos, asin, sqrt, ceil, pi
from typing import List


from models.meter import CandidateMeter, OutputMeter
from models.user import UserPreferences, Location


##### helper functions



#computes the distance between 2 lat/long points in meters using haversine 
def distance_meters(a: Location, b: Location) -> float:
    R = 6_371_000  # Earth radius in meters

    def to_rad(d: float) -> float:
        return d * pi / 180

    d_lat = to_rad(b.lat - a.lat)
    d_lon = to_rad(b.lon - a.lon)
    lat1 = to_rad(a.lat)
    lat2 = to_rad(b.lat)

    h = (
        sin(d_lat / 2) ** 2
        + cos(lat1) * cos(lat2) * sin(d_lon / 2) ** 2
    )

    return 2 * R * asin(sqrt(h))




#convers user preferences into scored weights 
def get_weights(prefs: UserPreferences):
    return {
        "distance": 0.45 if prefs.walking_distance_preference == "close" else 0.3,
        "cost": 0.45 if prefs.budget_range_preference == "low" else 0.3,
        "time": 0.25 if prefs.stay_time_preference == "long" else 0.2,
    }




##### main logic -- rank candidate parking spots based on scores. 

def rank_parking_spots(candidates: List[CandidateMeter],destination: Location, user_prefs: UserPreferences) -> List[OutputMeter]:

    weights = get_weights(user_prefs)

    MAX_DISTANCE = 1000
    MAX_COST = 20
    IDEAL_TIME = 180

    scored = []
    #computes each parking spot:
    for spot in candidates:
        distance = spot.distance_to_destination_meters
        walk_time = spot.walk_time_minutes
        estimated_total_cost = spot.estimated_total_cost

        # normalize scores
        distance_score = 1 - min(distance / MAX_DISTANCE, 1)
        cost_score = 1 - min(estimated_total_cost / MAX_COST, 1)
        time_score = min(spot.time_limit_minutes / IDEAL_TIME, 1)

        score = (
            weights["distance"] * distance_score
            + weights["cost"] * cost_score
            + weights["time"] * time_score
        )

        scored.append({
            "spot": spot,
            "score": score
        })

    # sort scores
    scored.sort(key=lambda x: x["score"], reverse=True)

    # assign ranks and return top K! -- for now hardcoded to top 8
    results: List[OutputMeter] = []

    for rank, (_, spot) in enumerate(scored[:8], start=1):
        result = OutputMeter( 
            spaceid=spot.spaceid,
            address=spot.address,
            rate_per_hour=spot.rate_per_hour,
            time_limit_minutes=spot.time_limit_minutes,
            occupancy=spot.occupancy,
            distance_to_destination_meters=spot.distance_to_destination_meters,
            walk_time_minutes=spot.walk_time_minutes,
            estimated_total_cost=spot.estimated_total_cost,
            rank=rank
        )

        results.append(result)

    return results